### DISCUSSION TITLE
"DockerClient" is not exported from module "python_on_whales"

### DISCUSSION THREAD

**Discussion by sylvainmouquet** [microsoft/pyright#3764] [Q&A]
Hello,

I have an error with pyright and with the usage of the library python_on_whales : 

```
error: "DockerClient" is not exported from module "python_on_whales"
error: Expected class type but received "DockerClient" (reportGeneralTypeIssues)
error: Expected class type but received "DockerClient"
```

the code is basic: 
```
from python_on_whales import DockerClient
from python_on_whales import docker as libdocker
```

I have already created an issue in the codebase of the project but it's not sure the problem is in the API.
https://github.com/gabrieldemarmiesse/python-on-whales/issues/356#issuecomment-1194434138

Could you confirm us, it's not an issue with pyright ?

Thanks

**Comment by erictraut** [comment]
Which version of `python-on-whales` are you using? I just installed the latest version (0.50.0), and I receive a different error than the one you've reported above. Perhaps you're using an older version?

I'm seeing the following error message:
```
"DockerClient" is not exported from module "python_on_whales" (reportPrivateImportUsage)
  Import from "python_on_whales.docker_client" instead
```

The `python-on-whales` library is marked as "py.typed" which means that it is asserting that it includes type information. When a library is "py.typed", pyright uses a set of established rules to determine which symbols are intended to be "public" and which are private. Refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-interface) for details.

The `ptyhon_on_whales/__init__.py` module is importing symbols like `DockerClient` from submodules. By default, imported symbols in a "py.typed" module are assumed to be private to that module's implementation, not publicly "re-exported" as public symbols from that module. If the intent is to re-export those symbols publicly (as it likely is in this case), the library authors would need to use a redundant form of import (e.g. `from from .docker_client import DockerClient as DockerClient`). This redundant form is specified in PEP 484 and was originally intended for use in type stub files, but it has since been adopted for use in "py.typed" modules as well.

Incidentally, the CLI version of pyright supports a special "--verifytypes" mode that can be used to validate the "type completeness" of a "py.typed" library. When I run `pyright --verifytypes python_on_whales` version 0.50.0, pyright produces the following summary:

```
Symbols exported by "python_on_whales": 2296
  With known type: 2026
  With ambiguous type: 49
  With unknown type: 221

Type completeness score: 88.2%
```

A completeness score of 88.2% is pretty good, but there is still some work to be done by the maintainers of this library before the type information is 100% complete.
